### PR TITLE
Feature/bngp 3746 4.1 metric file changes

### PR DESCRIPTION
**Pull request by soorajvdefra**
The changes from the tickets
https://eaflood.atlassian.net/browse/BNGP-3674
https://eaflood.atlassian.net/browse/BNGP-3438 
 are included in this Pull Request

### CODE REVIEW

File: packages/bng-metric-service/src/helpers/extractors/bng-metric-single-data-extractor.js [hunk 1]
```diff
@@ -110,30 +110,24 @@ class BngMetricSingleDataExtractor {
 
   #removeUnwantedColumns = (data, extractionConfiguration) => {
     const { columnsToBeRemoved, cellHeaders } = extractionConfiguration
+    const normalizedCellHeaders = cellHeaders.map(header => header.trim().toLowerCase())
```
Comment (paulsimonandrews) [inline packages/bng-metric-service/src/helpers/extractors/bng-metric-single-data-extractor.js:113]: The `.trim()` has changed the behaviour of this method, and I think caused you to change the tests to match the change.

I don't think the change here is right.

The metric seems to have a `' Habitat type'` column with a preceding whitespace. Before this method would have removed that column, but now it doesn't.

I don't think the issue happens with the config files use in the main code base as they have an `endCell` and I suspect `' Habitat type'` is a hidden column somewhere to the right of the `endCell`.

Preceding and trailing whitespace can already be handled by the `performSubstitution` method if needed.

I think you should revert to the original behaviour of this method.

**Review by paulsimonandrews — CHANGES_REQUESTED** [review]
Things to update:

1. You've changed the behaviour of the `removeUnwantedColumns` method, which I don't think is right. Tests have also been wrongly changed because of this.

2. You should update the payload in `webapp/src/utils/application.js` (and validation in `application-validation.js`) to include the Habitat reference as per the ACs of the ticket: https://eaflood.atlassian.net/browse/BNGP-3438 

3. If `'Habitat reference '` has a trailing whitespace you can use the `substitutions` functionality in the extraction config

4. Re add the `'Baseline ref'` checks to the tests that have been removed.

5. Update the test seed data to match the changes to payload and its validation (see point 2)

File: packages/bng-metric-service/src/helpers/extractors/__tests__/bng-metric-single-data-extractor.spec.js [hunk 2]
```diff
@@ -6,155 +6,166 @@ import areOffsiteTotalsCorrect from '../validation-config/are-offsite-totals-cor
...
-  it('should extract a valid metric v4', async () => {
-    const readableStreamv4 = fs.createReadStream('packages/bng-metric-service/src/__mock-data__/metric-file/metric-4.0.xlsm')
+  it('should extract a valid metric v4.1', async () => {
+    const readableStreamv4 = fs.createReadStream('packages/bng-metric-service/src/__mock-data__/metric-file/metric-4.1.xlsm')
     const bngMetricDataExtractor = new BngMetricSingleDataExtractor()
     const response = await bngMetricDataExtractor.extractContent(readableStreamv4, options)
 
-    expect(response.d1OffSiteHabitatBaseline.length).toEqual(5)
-    expect(Object.keys(response.d1OffSiteHabitatBaseline[0]).length).toEqual(9)
-    expect(response.d1OffSiteHabitatBaseline[0]['Baseline ref']).toEqual(1)
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/bng-metric-service/src/helpers/extractors/__tests__/bng-metric-single-data-extractor.spec.js:22]: updated tests

File: packages/bng-metric-service/src/helpers/extractors/__tests__/bng-metric-single-data-extractor.spec.js [hunk 3]
```diff
@@ -6,155 +6,166 @@ import areOffsiteTotalsCorrect from '../validation-config/are-offsite-totals-cor
...
-    expect((response.d2OffSiteHabitatCreation[0]['Habitat units delivered']).toFixed(2)).toEqual('7.01')
-    expect(response.d2OffSiteHabitatCreation[0]['Off-site reference']).toEqual(1234)
+    expect((response.d2OffSiteHabitatCreation[0]['Habitat units delivered']).toFixed(2)).toEqual('11.79')
+    expect(response.d2OffSiteHabitatCreation[0]['Habitat reference ']).toEqual('H1')
+    expect(response.d2OffSiteHabitatCreation[0]['Off-site reference']).toEqual('BNGREG-0001')
     expect(response.d2OffSiteHabitatCreation[0]['Strategic significance']).toEqual('Area/compensation not in local strategy/ no local strategy')
 
-    expect(response.d3OffSiteHabitatEnhancement.length).toEqual(4)
-    expect(Object.keys(response.d3OffSiteHabitatEnhancement[0]).length).toEqual(14)
-    expect(response.d3OffSiteHabitatEnhancement[0]['Baseline ref']).toEqual(2)
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/bng-metric-service/src/helpers/extractors/__tests__/bng-metric-single-data-extractor.spec.js:46]: updated tests

File: packages/bng-metric-service/src/helpers/extractors/__tests__/bng-metric-single-data-extractor.spec.js [hunk 4]
```diff
@@ -6,155 +6,166 @@ import areOffsiteTotalsCorrect from '../validation-config/are-offsite-totals-cor
...
-    expect((response.e2OffSiteHedgeCreation[0]['Hedge units delivered']).toFixed(2)).toEqual('1.77')
-    expect(response.e2OffSiteHedgeCreation[0]['Off-site reference']).toEqual(7156)
+    expect((response.e2OffSiteHedgeCreation[0]['Hedge units delivered']).toFixed(2)).toEqual('0.44')
+    expect(response.e2OffSiteHedgeCreation[0]['Habitat reference ']).toEqual('H1')
+    expect(response.e2OffSiteHedgeCreation[0]['Off-site reference']).toEqual('BNGREG-0001')
     expect(response.e2OffSiteHedgeCreation[0]['Strategic significance']).toEqual('Area/compensation not in local strategy/ no local strategy')
 
     expect(response.e3OffSiteHedgeEnhancement.length).toEqual(2)
-    expect(Object.keys(response.e3OffSiteHedgeEnhancement[0]).length).toEqual(11)
-    expect(response.e3OffSiteHedgeEnhancement[0]['Baseline ref']).toEqual(1)
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/bng-metric-service/src/helpers/extractors/__tests__/bng-metric-single-data-extractor.spec.js:85]: updated tests

File: packages/bng-metric-service/src/helpers/extractors/__tests__/bng-metric-single-data-extractor.spec.js [hunk 5]
```diff
@@ -6,155 +6,166 @@ import areOffsiteTotalsCorrect from '../validation-config/are-offsite-totals-cor
...
+    expect((response.d3OffSiteHabitatEnhancement[1]['Habitat units delivered']).toFixed(2)).toEqual('2.50')
+    expect((response.d3OffSiteHabitatEnhancement[1]['Total habitat area']).toFixed(2)).toEqual('0.37')
+    expect(response.d3OffSiteHabitatEnhancement[1]['Distinctiveness change']).toEqual('Medium - Medium')
+    expect(response.d3OffSiteHabitatEnhancement[1]['Condition change']).toEqual('Poor - Moderate')
+    expect(response.d3OffSiteHabitatEnhancement[1]['Habitat reference ']).toEqual('H1')
+    expect(response.d3OffSiteHabitatEnhancement[1]['Off-site reference']).toEqual(1234)
 
-    expect(response.e1OffSiteHedgeBaseline.length).toEqual(3)
-    expect(Object.keys(response.e1OffSiteHedgeBaseline[0]).length).toEqual(9)
-    expect(response.e1OffSiteHedgeBaseline[0]['Baseline ref']).toEqual(1)
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/bng-metric-service/src/helpers/extractors/__tests__/bng-metric-single-data-extractor.spec.js:62]: updated tests

File: packages/bng-metric-service/src/helpers/extractors/__tests__/bng-metric-single-data-extractor.spec.js [hunk 6]
```diff
@@ -6,155 +6,166 @@ import areOffsiteTotalsCorrect from '../validation-config/are-offsite-totals-cor
...
+    expect(response.e3OffSiteHedgeEnhancement[0]['Proposed habitat']).toEqual('Native hedgerow - associated with bank or ditch')
+    expect(response.e3OffSiteHedgeEnhancement[0]['Habitat reference']).toEqual('E3')
     expect(response.e3OffSiteHedgeEnhancement[0]['Off-site reference']).toEqual(1234)
-    expect(response.e3OffSiteHedgeEnhancement[0]['Strategic significance']).toEqual('Area/compensation not in local strategy/ no local strategy')
+    expect(response.e3OffSiteHedgeEnhancement[0]['Strategic significance']).toEqual('Location ecologically desirable but not in local strategy')
     expect(response.e3OffSiteHedgeEnhancement[0].Distinctiveness).toEqual('Medium')
 
-    expect(response.f1OffSiteWaterCBaseline.length).toEqual(3)
-    expect(Object.keys(response.f1OffSiteWaterCBaseline[0]).length).toEqual(10)
-    expect(response.f1OffSiteWaterCBaseline[0]['Baseline ref']).toEqual(1)
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/bng-metric-service/src/helpers/extractors/__tests__/bng-metric-single-data-extractor.spec.js:97]: updated tests

File: packages/bng-metric-service/src/helpers/extractors/__tests__/bng-metric-single-data-extractor.spec.js [hunk 7]
```diff
@@ -6,155 +6,166 @@ import areOffsiteTotalsCorrect from '../validation-config/are-offsite-totals-cor
...
+    expect(response.f2OffSiteWaterCCreation[0]['Extent of encroachment for both banks']).toEqual('Minor/ Minor')
     expect(response.f2OffSiteWaterCCreation[0]['Extent of encroachment']).toEqual('No Encroachment')
-    expect(response.f2OffSiteWaterCCreation[0]['Strategic significance']).toEqual('Formally identified in local strategy')
-    expect(response.f2OffSiteWaterCCreation[0]['Watercourse type']).toEqual('Ditches')
+    expect(response.f2OffSiteWaterCCreation[0]['Strategic significance']).toEqual('Area/compensation not in local strategy/ no local strategy')
+    expect(response.f2OffSiteWaterCCreation[0]['Watercourse type']).toEqual('Priority habitat')
 
     expect(response.f3OffSiteWaterCEnhancement.length).toEqual(2)
-    expect(Object.keys(response.f3OffSiteWaterCEnhancement[0]).length).toEqual(12)
-    expect(response.f3OffSiteWaterCEnhancement[0]['Baseline ref']).toEqual(1)
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/bng-metric-service/src/helpers/extractors/__tests__/bng-metric-single-data-extractor.spec.js:123]: updated tests

File: packages/bng-metric-service/src/helpers/extractors/__tests__/bng-metric-single-data-extractor.spec.js [hunk 8]
```diff
@@ -227,8 +239,13 @@ describe('BNG data extractor test', () => {
...
 
     const response = await bngMetricDataExtractor.extractContent(readableStreamv4, { extractionConfiguration })
-    expect(response.test.length).toEqual(4)
-    response.test.forEach(element => expect(Object.keys(element)).toEqual([baselineRef, habitatType]))
+    expect(response.test.length).toEqual(248)
+
+    response.test.forEach(element => {
+      const trimmedKeys = Object.keys(element).map(key => key.trim())
+      const habitatTypeCount = trimmedKeys.filter(key => key === habitatType).length
+      expect(habitatTypeCount).toBeGreaterThan(0)
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/bng-metric-service/src/helpers/extractors/__tests__/bng-metric-single-data-extractor.spec.js]: updated tests

File: packages/bng-metric-service/src/helpers/extractors/bng-metric-single-data-extractor.js [hunk 9]
```diff
@@ -110,30 +110,24 @@ class BngMetricSingleDataExtractor {
...
-        if (row[column]) {
-          delete row[column]
-        }
-      })
-
-      // Checking if all requested columns in configs are there
       Object.keys(row).forEach(key => {
-        if (!cellHeaders.includes(key)) {
+        const normalizedKey = key.trim().toLowerCase()
+        if (columnsToBeRemoved.includes(normalizedKey)) {
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/bng-metric-service/src/helpers/extractors/bng-metric-single-data-extractor.js]: i have removed this change and update the extraction config

File: packages/bng-metric-service/src/helpers/extractors/bng-metric-single-data-extractor.js (hunk 1 above)
Comment (soorajvdefra) [reply packages/bng-metric-service/src/helpers/extractors/bng-metric-single-data-extractor.js]: i have removed this change and update the extraction config

**Review by paulsimonandrews — CHANGES_REQUESTED** [review]
Small change required. The payload should have the habitat reference number added to `proposedHabitatId` for the proposed habitats. Not needed for the baselines.

File: packages/webapp/src/utils/application.js [hunk 10]
```diff
@@ -52,6 +53,7 @@ const getHabitats = session => {
       proposedHabitatId: '', // Leave blank for now until metric 4.1 when this property is extracted
       habitatType: details['Habitat type'] ?? details['Watercourse type'] ?? details['Proposed habitat'],
       baselineReference: details['Baseline ref'] ? String(details['Baseline ref']) : '',
+      habitatReferenceNumber: details['Habitat reference Number'] ?? details['Habitat reference Number'],
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/webapp/src/utils/application.js]: done

File: packages/webapp/src/utils/application.js [hunk 11]
```diff
@@ -38,6 +38,7 @@ const getHabitats = session => {
     metricData[identifier].filter(details => 'Baseline ref' in details).map(details => ({
       habitatType: details['Habitat type'] ?? details['Watercourse type'] ?? details['Hedgerow type'],
       baselineReference: String(details['Baseline ref']),
+      habitatReferenceNumber: details['Habitat reference Number'] ?? details['Habitat reference Number'],
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/webapp/src/utils/application.js]: done

**Review by paulsimonandrews — APPROVED** [review]

**Review by paulsimonandrews — APPROVED** [review]
LGTM

**Comment by shaundefra** [comment]
tested locally

**Pull request merged**